### PR TITLE
Adds a flavour only height system

### PR DESCRIPTION
**Pull request by FlamingLily**
Adds a new height field to the character creator, for height in cm. Adds flavour text to describe height differences between the examinee and examinor. Height values are restricted to limits, lore input will almost certainly be needed to verify that the inputs are correct and not too restrictive/freeing.

What this PR does NOT do:

- Sprite size changes.
- Mechanical benefits (Speed, health, whatever) to height.

ToDo:
- [X] Establish Height Framework and flavour text on examine
- [X] Establish height values in preferences and allow change on customisation
- [x] Fix the grammar of the flavour text when I'm not overtired
- [x] Ensure changelings mimic their target's height when changing their form
- [x] ~~Ensure antagonist character customiser can alter height values~~

> Decided against this one as my plan was just to add categories of heights, which would give a person an average height for whatever category. Figured this is literally no different than what would happen WITHOUT changing the customizer: Antagonists would default to their species average.

- [x] Get Lore approval of height restrictions

### CODE REVIEW

**Comment by Sniblet** [comment]
kornelija rigo is now officially too small to exist

**Comment by FlamingLily** [comment]
!review

**Comment by WhatsUpBrotendo** [comment]
Skrell height restrictions are fine.

**Comment by Eyeveri** [comment]
Is there any way you could add a alternate Ft/Inch measurement?

File: code/modules/mob/living/carbon/human/species/station/vaurca/vaurca_subspecies.dm [hunk 1]
```diff
@@ -51,6 +54,9 @@
 	name_plural = "Type CB"
 	bodytype = BODYTYPE_VAURCA_BREEDER
 	primitive_form = SPECIES_VAURCA_WARRIOR
+	species_height = HEIGHT_CLASS_GIGANTIC
+	height_min = 220
+	height_max = 320
```
Comment (FlamingLily) [inline code/modules/mob/living/carbon/human/species/station/vaurca/vaurca_subspecies.dm:59]:
```suggestion
	height_max = 335
```
bigger...

File: code/modules/mob/living/carbon/human/species/station/human/human.dm [hunk 2]
```diff
@@ -5,6 +5,9 @@
 	name_plural = "Humans"
 	category_name = "Human"
 	bodytype = BODYTYPE_HUMAN
+	species_height = HEIGHT_CLASS_AVERAGE
+	height_min = 120
+	height_max = 210
```
Comment (FlamingLily) [inline code/modules/mob/living/carbon/human/species/station/human/human.dm:10]:
```suggestion
	height_min = 140
	height_max = 203 //6'8", ugly number.
```

File: code/modules/mob/living/carbon/human/species/station/human/human_subspecies.dm [hunk 3]
```diff
@@ -4,6 +4,9 @@
 	blurb = "The Offworlders are humans that have adapted to zero-G conditions through a lifetime of conditioning, exposure, and physical modification. \
 	They thrive in thinner atmosphere and weightlessness, more often than not utilizing advanced life support and body-bracing equipment to sustain themselves in normal Human environments."
 	hide_name = FALSE
+	species_height = HEIGHT_CLASS_TALL
+	height_min = 150
```
Comment (FlamingLily) [inline code/modules/mob/living/carbon/human/species/station/human/human_subspecies.dm:8]:
```suggestion
	height_min = 180
```

File: code/modules/mob/living/carbon/human/species/station/ipc/ipc_subspecies.dm [hunk 4]
```diff
@@ -4,6 +4,9 @@
 	short_name = "jak"
 	name_plural = "Shells"
 	bodytype = BODYTYPE_HUMAN
+	species_height = HEIGHT_CLASS_AVERAGE
+	height_min = 120
```
Comment (FlamingLily) [inline code/modules/mob/living/carbon/human/species/station/ipc/ipc_subspecies.dm:8]:
```suggestion
	height_min = 140
```

**Review by FlamingLily — COMMENTED** [review]
Human heights via Schwann

File: code/modules/mob/mob.dm [hunk 5]
```diff
@@ -1472,3 +1472,56 @@
...
+			switch(height - selfCarbon.species.species_height)
+				if(-999 to -100)
+					descriptor = "miniscule"
+				if(-99 to -50)
+					descriptor = "tiny"
+				if(-49 to -20)
+					descriptor = "small"
+				if(-19 to 20)
+					descriptor = "about average height"
+				if(21 to 50)
```
Comment (FlamingLily) [inline code/modules/mob/mob.dm:1496]:
```suggestion
				if(-49 to -11)
					descriptor = "small"
				if(-10 to 10)
					descriptor = "about average height"
				if(11 to 50)
```

File: code/modules/mob/mob.dm [hunk 6]
```diff
@@ -1472,3 +1472,56 @@
...
+	switch(height - examiner.height)
+		if(-999 to -100)
+			descriptor = "absolutely tiny compared to"
+		if(-99 to -50)
+			descriptor = "much smaller than"
+		if(-49 to -20)
+			descriptor = "shorter than"
+		if(-19 to 20)
+			descriptor = "around about the same height"
+		if(21 to 50)
```
Comment (FlamingLily) [inline code/modules/mob/mob.dm:1517]:
```suggestion
		if(-49 to -11)
			descriptor = "shorter than"
		if(-10 to 10)
			descriptor = "around about the same height"
		if(11 to 50)
```

**Review by FlamingLily — COMMENTED** [review]
Wasn't happy with how distant the "about the same" height thing was so I narrowed it.

File: code/modules/mob/living/carbon/human/species/station/unathi/unathi.dm [hunk 7]
```diff
@@ -4,6 +4,9 @@
 	name_plural = "Unathi"
 	category_name = "Unathi"
 	bodytype = BODYTYPE_UNATHI
+	species_height = HEIGHT_CLASS_TALL
+	height_min = 170
+	height_max = 220
```
Comment (FlamingLily) [inline code/modules/mob/living/carbon/human/species/station/unathi/unathi.dm:9]:
```suggestion
	height_min = 175
	height_max = 215
```

**Comment by FlamingLily** [comment]
All heights confirmed by respective lore teams

File: code/game/atoms.dm [hunk 8]
```diff
@@ -244,6 +244,12 @@
 	to_chat(user, "[icon2html(src, user)] That's [f_name] [suffix]") // Object name. I.e. "This is an Object. It is a normal-sized item."
 	to_chat(user, desc)	// Object description.
 
+	// Height for Mobs -- Overwritten by examine proc for human; also called there
+	if(ismob(src))
+		var/mob/M = src
+		if(M.assembleHeightString(user))
+			to_chat(user, SPAN_NOTICE(M.assembleHeightString(user)))
```
Comment (NonQueueingMatt) [inline code/game/atoms.dm:251]: `istype(src, ...)` checks should be avoided. In this case, ismob(src).

File: code/modules/mob/mob.dm [hunk 9]
```diff
@@ -1472,3 +1472,56 @@
...
+
+//Height String for examine - Runs on the mob being examined.
+/mob/proc/assembleHeightString(mob/examiner)
+	var/heightString = null
+	var/descriptor
+	if(height == HEIGHT_NOT_USED)
+		return heightString
+
+	// Compare to Species Average
+	if(ishuman(src))
```
Comment (NonQueueingMatt) [inline code/modules/mob/mob.dm:1484]: As above, avoid `istype(src, ...) checks. This part can easily be a different proc that is overridden on different subtypes of /mob/.

**Review by NonQueueingMatt — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/human/species/station/human/human.dm [hunk 10]
```diff
@@ -5,6 +5,9 @@
 	name_plural = "Humans"
 	category_name = "Human"
 	bodytype = BODYTYPE_HUMAN
+	species_height = HEIGHT_CLASS_AVERAGE
+	height_min = 140
```
Comment (FlamingLily) [inline code/modules/mob/living/carbon/human/species/station/human/human.dm:9]:
```suggestion
	height_min = 145
```

File: code/modules/mob/living/carbon/human/species/station/ipc/ipc_subspecies.dm [hunk 11]
```diff
@@ -4,6 +4,9 @@
 	short_name = "jak"
 	name_plural = "Shells"
 	bodytype = BODYTYPE_HUMAN
+	species_height = HEIGHT_CLASS_AVERAGE
+	height_min = 140
+	height_max = 230 // If for whatever reason you wanted a shell that was ~~pretending to be geneboosted~~ SCHLOSSER
```
Comment (NonQueueingMatt) [inline code/modules/mob/living/carbon/human/species/station/ipc/ipc_subspecies.dm:9]:
```suggestion
	height_max = 230
```
No references to characters in the code, please.

File: SQL/migrate/V078__char_heights.sql [hunk 12]
```diff
@@ -0,0 +1,5 @@
+--
+-- Adds character heights
+--
+
+ALTER TABLE `ss13_characters` ADD COLUMN `char_height` INT(3) NOT NULL DEFAULT 0 AFTER `species`
```
Comment (Arrow768) [inline SQL/migrate/V078__char_heights.sql:5]: We are in the ss13_characters table, so the prefix char_ is superfluous.

File: code/modules/client/preferences.dm [hunk 13]
```diff
@@ -39,6 +39,7 @@ datum/preferences
 	var/gender = MALE					//gender of character (well duh)
 	var/pronouns = NEUTER				//what the character will appear as to others when examined
 	var/age = 30						//age of character
+	var/char_height						//character's height
```
Comment (Arrow768) [inline code/modules/client/preferences.dm:42]: Same here.
We are already in the char preferences.

**Review by Arrow768 — CHANGES_REQUESTED** [review]

**Review by Arrow768 — APPROVED** [review]

**Review by NonQueueingMatt — APPROVED** [review]

**Pull request merged**